core/drop_test.go

```diff
@@ -1,19 +1,222 @@
+						buildPrePrepareMessageFn:  node.buildPrePrepare,
+						buildPrepareMessageFn:     node.buildPrepare,
+						buildCommitMessageFn:      node.buildCommit,
+						buildRoundChangeMessageFn: node.buildRoundChange,
+
+						insertBlockFn: func(proposal []byte, _ []*messages.CommittedSeal) {
+							insertedBlocks[i] = proposal
+						},
+						hasQuorumFn: c.hasQuorumFn,
+					},
+					&mockTransport{multicastFn: func(message *proto.Message) {
+						if currentNode.offline {
+							return
+						}
+
+						c.gossip(message)
+					}},
+				)
+			}
+		},
+	)
+
+	// Progress the chain to claim it works ok by default
+	err := cluster.progressToHeight(5*time.Second, 1)
+	assert.NoError(t, err, "unable to reach height: %w", err)
+	assert.Equal(t, uint64(1), cluster.latestHeight)
+	assertValidInsertedBlocks(insertedBlocks, t) // Make sure the inserted blocks are valid
+
+	insertedBlocks = make([][]byte, numNodes) // Purge
+
+	// Stop all nodes and make sure no blocks are written
+	cluster.stopN(len(cluster.nodes))
+	cluster.progressToHeight(2*time.Second, 2)
+	assertNoInsertedBlocks(insertedBlocks, t)
+
+	// Start all and expect valid blocks to be written again
+	cluster.startN(len(cluster.nodes))
+	cluster.progressToHeight(5*time.Second, 10)
+	assertValidInsertedBlocks(insertedBlocks, t) // Make sure the inserted blocks are valid
+}
+
+func assertNoInsertedBlocks(blocks [][]byte, t *testing.T) {
+	t.Helper()
+
+	assertNInsertedBlocks(0, blocks, t)
+}
+
+func assertNInsertedBlocks(n int, blocks [][]byte, t *testing.T) {
+	t.Helper()
+
+	writtenBlocks := 0
+	for _, block := range blocks {
+		if !bytes.Equal(block, nil) {
+			writtenBlocks++
+		}
+	}
+
+	assert.True(t, n == writtenBlocks)
+}
+
+func assertValidInsertedBlocks(blocks [][]byte, t *testing.T) {
+	t.Helper()
+
+	for _, block := range blocks {
+		assert.True(t, bytes.Equal(block, validProposal))
+	}
+}
+
+func TestMaxFaultyPlusOneDroppingMessages(t *testing.T) {
+	t.Parallel()
+
+	cluster := newCluster(
+		6,
+		func(c *cluster) {
+			for _, node := range c.nodes {
+				currentNode := node
+				node.core = NewIBFT(
+					mockLogger{},
+					&mockBackend{
+						isValidBlockFn:         isValidProposal,
+						isValidProposalHashFn:  isValidProposalHash,
+						isValidSenderFn:        nil,
+						isValidCommittedSealFn: nil,
+						isProposerFn:           c.isProposer,
+
+						idFn: node.addr,
+
+						buildProposalFn:           buildValidProposal,
+						buildPrePrepareMessageFn:  node.buildPrePrepare,
+						buildPrepareMessageFn:     node.buildPrepare,
+						buildCommitMessageFn:      node.buildCommit,
+						buildRoundChangeMessageFn: node.buildRoundChange,
+
+						insertBlockFn: nil,
+						hasQuorumFn:   c.hasQuorumFn,
+					},
+					&mockTransport{multicastFn: func(message *proto.Message) {
+						if currentNode.faulty && rand.Intn(100) < 50 {
+							return
+						}
+
+						c.gossip(message)
+					}},
+				)
+			}
+		},
+	)
+
+	cluster.makeNFaulty(int(cluster.maxFaulty()))
```

Comment:
Is this MaxFaultyPlusOne?

Reply:
Yes i renamed the test since the max faulty plus one can actually pass as a test